x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
+
+                Map<String, NodeStats> nodeStatsMap = nodesStatsResponse.getNodesMap();
+                for (String nodeId : allNodeIds) {
+                    if (nodeStatsMap.get(nodeId) != null) {
+                        List<IndexShardStats> indexShardStatsList = nodeStatsMap.get(nodeId).getIndices().
+                            getShardStats(indexMetadata.getIndex());
+                        long shardsOnCurrentNodeStorageBytes = indexShardStatsList.stream().mapToLong(indexShardStats ->
+                            Arrays.stream(indexShardStats.getShards()).mapToLong(shardStats ->
+                                shardStats.getStats().getStore().getSizeInBytes()).sum()).sum();
+                        if (shardsOnCurrentNodeStorageBytes != 0) {
+                            nodeShardsStorageBytes.put(nodeId, shardsOnCurrentNodeStorageBytes);
+                        }
+                    }
+                }
+                long indexPrimaryShardsStorageBytes = nodeShardsStorageBytes.values().stream().mapToLong(Long::longValue).sum();
+                if (indexMetadata.getNumberOfReplicas() != 0) {
+                    indexPrimaryShardsStorageBytes /= indexMetadata.getNumberOfReplicas();
+                }
+                for (RoutingNode node : routingNodes) {
+                    if (nodeStatsMap.get(node.nodeId()) != null) {
+                        FsInfo.Path totalFsInfo = nodeStatsMap.get(node.nodeId()).getFs().getTotal();
+                        long nodeTotalBytes = totalFsInfo.getTotal().getBytes();
+                        long nodeAvailableBytes = totalFsInfo.getAvailable().getBytes();
+                        long freeBytesThresholdLow;
+                        if (diskThresholdSettings.getFreeDiskThresholdLow() != 0) {
+                            freeBytesThresholdLow = (long) Math.ceil(nodeTotalBytes *
+                                diskThresholdSettings.getFreeDiskThresholdLow() * 0.01);
+                        } else {
+                            freeBytesThresholdLow = diskThresholdSettings.getFreeBytesThresholdLow().getBytes();
+                        }
+                        // we should make sure that allocating two copies of the index's primary shards to that node doesn't put the node
+                        // above the low watermark, if not, the new shrunken index may not be initialized successfully
+                        long shardsOnCurrentNodeStorageBytes = 0;
+                        if (nodeShardsStorageBytes.containsKey(node.nodeId())) {
+                            shardsOnCurrentNodeStorageBytes = nodeShardsStorageBytes.get(node.nodeId());
+                        }
+                        if (nodeAvailableBytes > freeBytesThresholdLow + 2 * indexPrimaryShardsStorageBytes -
+                            shardsOnCurrentNodeStorageBytes) {
+                            validRoutingNodes.add(node);
+                        }
```

Comment:
I think we'll need some sort of signalling here when there isn't enough space, so potentially capturing the event where none of the otherwise valid nodes can hold all the shards, and changing the Exception thrown in that case to include that message.